examples/keras_recipes/parameter_efficient_finetuning_of_gemma_with_lora_and_qlora.py

```diff
@@ -0,0 +1,365 @@
+"""
+Title: Parameter-efficient fine-tuning of Gemma with LoRA and QLoRA
+Authors: [Hongyu Chiu](https://github.com/james77777778), [Abheesht Sharma](https://github.com/abheesht17/), [Matthew Watson](https://github.com/mattdangerw/)
+Date created: 2024/07/12
+Last modified: 2024/07/12
+Description: Use KerasNLP to fine-tune a Gemma LLM with LoRA and QLoRA.
+Accelerator: GPU
+"""
+
+"""
+## Introduction
+
+Large Language Models (LLMs) have been shown to be effective at a variety of NLP
+tasks. An LLM is first pre-trained on a large corpus of text in a
+self-supervised fashion. Pre-training helps LLMs learn general-purpose
+knowledge, such as statistical relationships between words. An LLM can then be
+fine-tuned on a downstream task of interest (such as sentiment analysis).
+
+However, LLMs are extremely large in size, and we don't need to train all the
+parameters in the model while fine-tuning, especially because datasets on which
+the model is fine-tuned are relatively small. Another way of saying this is
+that LLMs are over-parametrized for fine-tuning. This is where
+[Low-Rank Adaptation (LoRA)](https://arxiv.org/abs/2106.09685) comes in; it
+significantly reduces the number of trainable parameters. This results in a
+decrease in training time and GPU memory usage, while maintaining the quality
+of the outputs.
+
+Furthermore,
+[Quantized Low-Rank Adaptation (QLoRA)](https://arxiv.org/abs/2305.14314)
+extends LoRA to enhance efficiency through quantization techniques without
+performance degradation.
+
+In this example, we will fine-tune KerasNLP's
+[Gemma model](https://keras.io/api/keras_nlp/models/gemma/) on the next token
+prediction task using LoRA and QLoRA.
+
+Note that this example runs on all backends supported by Keras. TensorFlow is
+only used for data preprocessing.
+"""
+
+"""
+## Setup
+
+Before we start implementing the pipeline, let's install and import all the
+libraries we need. We'll be using the KerasNLP library.
+
+Secondly, let's set the precision to bfloat16. This will help us reduce the
+memory usage and training time.
+
+Also, ensure that `KAGGLE_USERNAME` and `KAGGLE_KEY` have been correctly
+configured to access the Gemma model.
+"""
+
+"""shell
+pip install -q git+https://github.com/keras-team/keras-nlp.git
+pip install -q --upgrade keras
+"""
+
+import os
+
+os.environ["KERAS_BACKEND"] = "tensorflow"
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
```

Comment:
Why is this necessary?

Reply:
`TF_CPP_MIN_LOG_LEVEL=3` is used to suppress verbose logging from TF